### DISCUSSION TITLE
Deprecating Node.js 12 - strategy

### DISCUSSION THREAD

**Discussion by mrtnzlml** [adeira/universe#1588] [General]
Node.js 12 is currently "maintenance LTS" (https://nodejs.org/en/about/releases/) and instead, there is Node.js 14 (active LTS). There is also Node.js 15 (current) and Node.js 16 coming soon.

At this moment, we actively support versions 12, 14, and 15. According to Node.js recommendations, production applications should only use _active LTS_ or _maintenance LTS_ releases. This requirement would be fulfilled even with the removal of Node.js 12. So, I'd like to propose removing Node.js 12 and supporting only versions 14, 15 (and 16 when it's ready).

## Questions

How should we do the deprecation? We can safely ignore packages with `0.x`, however, other packages would need to be released with a major bump. Is this the right way to go? How would you do it? What should we write to the changelog?

Does anyone depend on Node.js 12 and require extra time for the migration?

## Proposal

1. write the following paragraph into `CHANGELOG.md` files of **all packages** (please help me with the wording):

> Support for Node.js 12 has been removed. This package might continue working on older Node.js versions, however, it's highly recommended to upgrade to Node.js version 14 or newer. For more details, see: https://nodejs.org/en/about/releases/

2. major bump all packages with version >=1.0.0 (might be unnecessary in some cases)
3. remove Node.js 12 from CI (once every package with version >= 1.0.0 is released)

## Alternatives

- Wait until Node.js version 16 and support 12 until then and follow the same process as described above?
- Wait until the end-of-life date of Node.js 12? (2022-04-30)

**Comment by michalsanger** [comment]
Shouldn’t we wait until Vercel supports v14? They are blocked by AWS Lambda https://github.com/vercel/vercel/discussions/4754

**Reply by mrtnzlml** [reply]
Good point, I didn't know this is a thing.

**Comment by tbergquist-godaddy** [comment]
I think the proposed solution sounds good, but I also agree with @michalsanger that we should wait until vercel supports v14, since we have code running there 🙂

**Comment by michalsanger** [comment]
Vercel supports v14 now: https://vercel.com/changelog/node-js-14-lts-is-now-available

**Reply by mrtnzlml** [reply]
Cool. So can we proceed with this? Do you agree with the drafted migration proposal @michalsanger?

**Comment by mrtnzlml** [comment]
Done: https://github.com/adeira/universe/issues/1718
